partcad/src/partcad/project_factory_git.py

```diff
@@ -107,7 +108,9 @@ def _clone_or_update_repo(self, repo_url, cache_dir=None):
                             branch_name = remote_head.reference.name
                             short_branch_name = branch_name[branch_name.find("/") + 1 :]
                             pc_logging.debug("Refreshing the GIT branch: %s" % short_branch_name)
-                            origin.pull(short_branch_name)
+                            origin.pull(short_branch_name, env={
+                                "GIT_CONFIG_PARAMETERS": git_config_params
```

Comment:
How this was validated?

For example doing verbose install does not show actual `git clone ...` commands:
```
pc -v --no-ansi init
pc -v --no-ansi install
```
For example if I look on repos configs they will still have `https:` even if it was cloned over SSH.

```
⬢ [Docker] ❯ find ~/.partcad/git -path */.git/config -exec grep url {} \;
        url = https://github.com/partcad/partcad-index.git
        url = https://github.com/openvmp/partcad-standard-metric-nema/
...
```

but if I break SSH Agent

```bash
unset SSH_AUTH_SOCK
```

Then it fails as expected:

```
⬢ [Docker] ❯ pc install
[ 0 / 2 ] InitCtx /tmp/PCO-14; 0s
Traceback (most recent call last):
  File "/workspaces/partcad/partcad/src/partcad/project_factory_git.py", line 162, in _clone_or_update_repo
    repo = Repo.clone_from(repo_url, cache_path, env={
  File "/workspaces/partcad/.venv/lib/python3.10/site-packages/git/repo/base.py", line 1525, in clone_from
    return cls._clone(
  File "/workspaces/partcad/.venv/lib/python3.10/site-packages/git/repo/base.py", line 1396, in _clone
    finalize_process(proc, stderr=stderr)
  File "/workspaces/partcad/.venv/lib/python3.10/site-packages/git/util.py", line 504, in finalize_process
    proc.wait(**kwargs)
  File "/workspaces/partcad/.venv/lib/python3.10/site-packages/git/cmd.py", line 834, in wait
    raise GitCommandError(remove_password_if_present(self.args), status, errstr)
git.exc.GitCommandError: Cmd('git') failed due to: exit code(128)
  cmdline: git clone -v -- https://github.com/partcad/partcad-index.git /home/vscode/.partcad/git/74d482ac304a898f1276960c537dae75cb6208a91b2086097ad1509761785e81
  stderr: 'Cloning into '/home/vscode/.partcad/git/74d482ac304a898f1276960c537dae75cb6208a91b2086097ad1509761785e81'...
remote username contains invalid characters
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/workspaces/partcad/.venv/bin/pc", line 6, in <module>
    sys.exit(cli())
  File "/workspaces/partcad/.venv/lib/python3.10/site-packages/rich_click/rich_command.py", line 367, in __call__
    return super().__call__(*args, **kwargs)
  File "/workspaces/partcad/.venv/lib/python3.10/site-packages/click/core.py", line 1161, in __call__
    return self.main(*args, **kwargs)
  File "/workspaces/partcad/.venv/lib/python3.10/site-packages/rich_click/rich_command.py", line 152, in main
    rv = self.invoke(ctx)
  File "/workspaces/partcad/.venv/lib/python3.10/site-packages/click/core.py", line 1697, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/workspaces/partcad/.venv/lib/python3.10/site-packages/click/core.py", line 1443, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/workspaces/partcad/.venv/lib/python3.10/site-packages/click/core.py", line 788, in invoke
    return __callback(*args, **kwargs)
  File "/workspaces/partcad/.venv/lib/python3.10/site-packages/click/decorators.py", line 45, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "<string>", line 13, in cli
  File "/workspaces/partcad/partcad/src/partcad/context.py", line 425, in get_all_packages
    self.import_all(parent_name)
  File "/workspaces/partcad/partcad/src/partcad/context.py", line 318, in import_all
    asyncio.run(self._import_all_wrapper(self.projects[parent_name]))
  File "/usr/local/python/3.10.16/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/python/3.10.16/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/workspaces/partcad/partcad/src/partcad/context.py", line 338, in _import_all_wrapper
    next_project = import_task.result()
  File "/workspaces/partcad/partcad/src/partcad/sync_threads.py", line 45, in run
    return await asyncio.get_running_loop().run_in_executor(executor, method, *args)
  File "/usr/local/python/3.10.16/lib/python3.10/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/workspaces/partcad/partcad/src/partcad/context.py", line 167, in import_project
    rfg.ProjectFactoryGit(self, parent, project_import_config)
  File "/workspaces/partcad/partcad/src/partcad/project_factory_git.py", line 47, in __init__
    self.path = self._clone_or_update_repo(self.import_config_url)
  File "/workspaces/partcad/partcad/src/partcad/project_factory_git.py", line 174, in _clone_or_update_repo
    raise RuntimeError(f"Failed to clone a repo: {e}")
RuntimeError: Failed to clone a repo: Cmd('git') failed due to: exit code(128)
  cmdline: git clone -v -- https://github.com/partcad/partcad-index.git /home/vscode/.partcad/git/74d482ac304a898f1276960c537dae75cb6208a91b2086097ad1509761785e81
  stderr: 'Cloning into '/home/vscode/.partcad/git/74d482ac304a898f1276960c537dae75cb6208a91b2086097ad1509761785e81'...
remote username contains invalid characters
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
INFO:  DONE: InitCtx: /tmp/PCO-14: 0.00s
INFO:  Cloning the GIT repo: https://github.com/partcad/partcad-index.git
INFO:  DONE: Install: this: 0.02s
```

so it does seem to work, but I do not see other way to validate it for now.

Reply:
How about updating import_config_url, using a function like:

![image](https://github.com/user-attachments/assets/7607957c-f042-4bbd-a6b6-aeedb8f40c5d)

which will result in:

![image](https://github.com/user-attachments/assets/c5824ab1-3a3a-43a1-9d12-9dda42c1c739)

there is no need to set "GIT_CONFIG_PARAMETERS" as git will automatically use ssh if url is of the form `git@github.com:`

so after cloning if you look on repos configs they will have `git@github.com:`

![image](https://github.com/user-attachments/assets/6759a77f-25ec-4a24-b900-a8943187f329)

Reply:
Overall I like this, but if you go with implementation in `import_config_url` then probably it could use simpler and more specific setting section in config, for example:

```yaml
import:
  overrides:
    url:
      "git@github.com:": "https://github.com/"
git:
  config:
    "user.name": "Agent Smith"
```

IMHO it would be nice to keep the `config.yaml:git.config[]` feature. It could prove useful in future and I like your suggestion where actual `url` is saved to `.git/config` for pulled repos.

P. S. For all new configurations added in config please add corresponding sections with examples in `docs/source/...`. Thanks!